Improvement/feature request: consider allowing process under test to timeout BEFORE Bazel-level `--test_timeout`

Under normal circumstances it is highly desirable to let the binary under test in a `go_test` rule time out *before Bazel allows the test to timeout*. This is because when the binary under test times out (e.g. using `-test.timeout`), it dumps stacktraces and some other potentially useful info.

In `cockroach` we use [this commit](https://github.com/cockroachdb/rules_go/commit/1360c0a3a4a1b3ff12ed685013e34b37f99fdf01) which checks the `TEST_TIMEOUT` and, when spawning the test subprocess, passes `-test.timeout N` where `N` is 3 seconds *less* than the `TEST_TIMEOUT`. (The 3 second figure is selected arbitrarily and is meant to give "enough time" for the binary to clean up.) This means that if you run `bazel test :test_target --test_timeout 60`, then the *actual* timeout will be 57 seconds (as 60 - 3 = 57).

While the disparity between the `go test`-level timeout and the Bazel-level timeout may be confusing at first, it creates a hugely superior UX. Compare the `test.log` output from a timed-out test without that patch:

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //pkg/cmd/dev:dev_test
-----------------------------------------------------------------------------
=== RUN   TestDataDriven
-- Test timed out at 2023-11-07 17:49:38 UTC --
```

This says almost nothing except that a timeout occurred, and the test is not given any time to clean up before being unceremoniously killed. Compare that to *with* the patch:

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //pkg/cmd/dev:dev_test
-----------------------------------------------------------------------------
=== RUN   TestDataDriven
panic: test timed out after 2s
running tests:
	TestDataDriven (2s)

goroutine 7 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2259 +0x320
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38

goroutine 1 [chan receive]:
testing.(*T).Run(0x140000a7380, {0x1051300c0?, 0x66f0a118049?}, 0x10531a668)
...
```

Now we get stack traces.

This change carries two "regressions" in the UX:
1. The timeout you pass to `--timeout` is not the real timeout, and the "real" timeout is a few seconds shorter. In `cockroach` we have gotten used to this and factor it in, though the behavior could be confusing for others.
2. Timed-out tests show as `FAILED` instead of `TIMEOUT` in the Bazel-produced test summary which may be technically inaccurate.

For my taste, neither of these "regressions" outweighs the benefit gained from allowing the test process time to gracefully clean up, report stack traces, etc.

### What version of rules_go are you using?

`release-0.42`

### What version of gazelle are you using?

`v0.33.0`

### What version of Bazel are you using?

6.2.1

### Does this issue reproduce with the latest releases of all the above?

Yes

### What operating system and processor architecture are you using?

MacOS/ARM64

### Any other potentially useful information about your toolchain?

No

--- a/go/tools/builders/generate_test_main.go
+++ b/go/tools/builders/generate_test_main.go
@@ -235,7 +235,7 @@ func main() {
 		}
 	}
 	{{end}}
-
+	bzltestutil.RegisterTimeoutHandler()
 	{{if not .TestMain}}
 	res := m.Run()
 	{{else}}

--- a/go/tools/bzltestutil/BUILD.bazel
+++ b/go/tools/bzltestutil/BUILD.bazel
@@ -5,6 +5,7 @@ go_tool_library(
     srcs = [
         "lcov.go",
         "test2json.go",
+        "timeout.go",
         "wrap.go",
         "xml.go",
     ],

--- a/go/tools/bzltestutil/timeout.go
+++ b/go/tools/bzltestutil/timeout.go
@@ -1,0 +1,36 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bzltestutil
+
+import (
+	"os"
+	"os/signal"
+	"runtime/debug"
+	"syscall"
+)
+
+func RegisterTimeoutHandler() {
+	go func() {
+		// When the Bazel test timeout is reached, Bazel sends a SIGTERM. We
+		// panic just like native go test would so that the user gets stack
+		// traces of all running go routines.
+		// See https://github.com/golang/go/blob/e816eb50140841c524fd07ecb4eaa078954eb47c/src/testing/testing.go#L2351
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, syscall.SIGTERM)
+		<-c
+		debug.SetTraceback("all")
+		panic("test timed out")
+	}()
+}

--- a/go/tools/bzltestutil/wrap.go
+++ b/go/tools/bzltestutil/wrap.go
@@ -23,10 +23,12 @@ import (
 	"log"
 	"os"
 	"os/exec"
+	"os/signal"
 	"path/filepath"
 	"strconv"
 	"strings"
 	"sync"
+	"syscall"
 
 	"github.com/bazelbuild/rules_go/go/tools/bzltestutil/chdir"
 )
@@ -67,9 +69,9 @@ func shouldAddTestV() bool {
 // ensure that one line at a time is written to the inner writer.
 type streamMerger struct {
 	OutW, ErrW *io.PipeWriter
-	mutex sync.Mutex
-	inner io.Writer
-	wg sync.WaitGroup
+	mutex      sync.Mutex
+	inner      io.Writer
+	wg         sync.WaitGroup
 	outR, errR *bufio.Reader
 }
 
@@ -78,10 +80,10 @@ func NewStreamMerger(w io.Writer) *streamMerger {
 	errR, errW := io.Pipe()
 	return &streamMerger{
 		inner: w,
-		OutW: outW,
-		ErrW: errW,
-		outR: bufio.NewReader(outR),
-		errR: bufio.NewReader(errR),
+		OutW:  outW,
+		ErrW:  errW,
+		outR:  bufio.NewReader(outR),
+		errR:  bufio.NewReader(errR),
 	}
 }
 
@@ -126,6 +128,17 @@ func Wrap(pkg string) error {
 	cmd.Env = append(os.Environ(), "GO_TEST_WRAP=0")
 	cmd.Stderr = io.MultiWriter(os.Stderr, streamMerger.ErrW)
 	cmd.Stdout = io.MultiWriter(os.Stdout, streamMerger.OutW)
+	go func() {
+		// When the Bazel test timeout is reached, Bazel sends a SIGTERM that
+		// we need to forward to the inner process.
+		// TODO: This never triggers on Windows, even though Go should simulate
+		//  a SIGTERM when Windows asks the process to close. It is not clear
+		//  whether Bazel uses the required graceful shutdown mechanism.
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, syscall.SIGTERM)
+		<-c
+		cmd.Process.Signal(syscall.SIGTERM)
+	}()
 	streamMerger.Start()
 	err := cmd.Run()
 	streamMerger.ErrW.Close()

--- a/tests/core/go_test/BUILD.bazel
+++ b/tests/core/go_test/BUILD.bazel
@@ -264,3 +264,8 @@ go_bazel_test(
     name = "filter_test_cases_test",
     srcs = ["filter_test_cases_test.go"],
 )
+
+go_bazel_test(
+    name = "timeout_test",
+    srcs = ["timeout_test.go"],
+)

--- a/tests/core/go_test/timeout_test.go
+++ b/tests/core/go_test/timeout_test.go
@@ -1,0 +1,66 @@
+package timeout_test
+
+import (
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"testing"
+
+	"github.com/bazelbuild/rules_go/go/tools/bazel_testing"
+)
+
+func TestMain(m *testing.M) {
+	bazel_testing.TestMain(m, bazel_testing.Args{
+		Main: `
+-- BUILD.bazel --
+load("@io_bazel_rules_go//go:def.bzl", "go_test")
+
+go_test(
+	name = "timeout_test",
+	srcs = ["timeout_test.go"],
+)
+-- timeout_test.go --
+package timeout
+
+import "testing"
+
+func TestFoo(t *testing.T) {
+	neverTerminates()
+}
+
+func neverTerminates() {
+	for {}
+}
+`,
+	})
+}
+
+func TestTimeout(t *testing.T) {
+	if runtime.GOOS == "windows" {
+		t.Skip("panics on timeouts are not yet supported on Windows")
+	}
+
+	if err := bazel_testing.RunBazel("test", "//:timeout_test", "--test_timeout=3"); err == nil {
+		t.Fatal("expected bazel test to fail")
+	} else if exitErr, ok := err.(*bazel_testing.StderrExitError); !ok || exitErr.Err.ExitCode() != 3 {
+		t.Fatalf("expected bazel test to fail with exit code 3", err)
+	}
+	p, err := bazel_testing.BazelOutput("info", "bazel-testlogs")
+	if err != nil {
+		t.Fatalf("could not find testlogs root: %s", err)
+	}
+	path := filepath.Join(strings.TrimSpace(string(p)), "timeout_test/test.log")
+	b, err := os.ReadFile(path)
+	if err != nil {
+		t.Fatalf("could not read test log: %s", err)
+	}
+
+	testLog := string(b)
+	if !strings.Contains(testLog, "panic: test timed out") {
+		t.Fatalf("test log does not contain expected panic:\n%s", testLog)
+	}
+	if !strings.Contains(testLog, "timeout_test.neverTerminates(") {
+		t.Fatalf("test log does not contain expected stack trace:\n%s", testLog)
+	}
+}

